modules/diskq/tests/test_diskq_truncate.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2018 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "logqueue.h"
+#include "logqueue-fifo.h"
+#include "logqueue-disk.h"
+#include "logqueue-disk-reliable.h"
+#include "logqueue-disk-non-reliable.h"
+#include "diskq.h"
+#include "qdisk.c"
+#include "logpipe.h"
+#include "apphook.h"
+#include "plugin.h"
+#include "mainloop.h"
+#include "mainloop-call.h"
+#include "mainloop-io-worker.h"
+
+#include "queue_utils_lib.h"
+#include "test_diskq_tools.h"
+#include "testutils.h"
+
+#include <stdlib.h>
+#include <string.h>
+#include <iv.h>
+#include <sys/stat.h>
+
+
+MsgFormatOptions parse_options;
+
+static LogQueue *
+_get_diskqueue(gchar *filename, DiskQueueOptions *options)
+{
+  LogQueue *q = log_queue_disk_non_reliable_new(options, NULL);
+  log_queue_set_use_backlog(q, FALSE);
+  log_queue_disk_load_queue(q, filename);
+  return q;
+}
+
+static LogQueue *
+_create_diskqueue(gchar *filename, DiskQueueOptions *options)
+{
+  LogQueue *q;
+  unlink(filename);
+
+  _construct_options(options, 1100000, 0, FALSE);
+  options->qout_size = 40;
+  q = _get_diskqueue(filename, options);
+  return q;
+}
+
+static void
+_save_diskqueue(LogQueue *q)
+{
+  gboolean persistent;
+  log_queue_disk_save_queue(q, &persistent);
+  log_queue_unref(q);
+}
+
+typedef struct
+{
+  gint number_of_msgs_to_push;
+  gint number_of_msgs_to_pop;
+  gboolean should_file_be_empty_after_truncate;
+} truncate_test_params;
+
+static void
+_test_diskq_truncate(truncate_test_params params)
+{
+  LogQueue *q;
+  DiskQueueOptions options;
+  GString *filename = g_string_new("test_dq_truncate.qf");
+
+  q = _create_diskqueue(filename->str, &options);
+  assert_gint(log_queue_get_length(q), 0, "Invalid queue length!");
+
+  feed_some_messages(q, params.number_of_msgs_to_push, &parse_options);
+  assert_gint(log_queue_get_length(q), params.number_of_msgs_to_push, "Invalid queue length!");
+  gint64 size_after_push = ((LogQueueDisk *)q)->qdisk->file_size;
+  struct stat after_push;
+  stat(filename->str, &after_push);
```

Comment:
I think we should not assert every step in a testcase:
- save_queue is not asserted
- unlink is not asserted either currently.

It is a good question what happens if stat failed (however I think it will not, as previous function would fail too).
I guess we would access uninitialized members inside the "struct stat".
Okay, it's not a major thing, I assert the stat() calls too.